### DISCUSSION TITLE
How to specify table schema on PostgreSQL other than public?

### DISCUSSION THREAD

**Discussion by Abdillah** [diesel-rs/diesel#3806] [Q&A] [answered]
When specifying data source for a struct,

```
#[derive(Queryable, Selectable, Debug)]
#[diesel(table_name = activity_history)]
struct Activity {
    // ....
}
```

we can't specify a full table name like `tracker.activity_history` into the proc macro.
Then, how to specify [PostgreSQL table schema](https://www.postgresql.org/docs/current/ddl-schemas.html) in diesel?

---
I have tried passing connection string `currentSchema=tracker` or `search_path=tracker` but it said unsupported. I use [Neon PostgreSQL](https://neon.tech/) if somebody know the correct param.

**Comment by weiznich** [comment]
It's not possible to answer this without knowing the relevant context. Please at at least the following details to your questions:

* Information about your diesel version
* Information about your rustc version
* Relevant parts of your schema, code, …

**Reply by Abdillah** [reply]
I'm starting new project, so I'm happy with any version of `rustc` and `diesel`.
Let's say I'm going for the latest version:

- Rust v1.72.1 (latest)
- Diesel v2.1.2 (latest)

I'm moving away from diesel on this project, so I will try to reconstruct my case for the sake of future myself.

My schema consists only one table.

```rust
diesel::table! {
    activity_history (id) {
        id -> BigInt,
        timestamp -> Timestamp,
        reporter -> Text,
        activity -> Text,
        data -> Json,
    }
}

#[derive(Queryable, Selectable, Debug)]
#[diesel(table_name = activity_history)]
struct Activity {
    pub id: i64,
    pub timestamp: chrono::NaiveDateTime,
    pub reporter: String,
    pub activity: String,
    pub data: serde_json::Value,
}
```

I setup diesel connection by using diesel_async (thanks @weiznich) and rustls.
This code is exactly as in the example somewhere (likely diesel_async's).

```rust
fn establish_connection(
    config: &str,
) -> BoxFuture<
    ConnectionResult<
        diesel_async::async_connection_wrapper::AsyncConnectionWrapper<AsyncPgConnection>,
    >,
> {
    let fut = async {
        // We first set up the way we want rustls to work.
        let rustls_config = rustls::ClientConfig::builder()
            .with_safe_defaults()
            .with_root_certificates(root_certs())
            .with_no_client_auth();
        let tls = tokio_postgres_rustls::MakeRustlsConnect::new(rustls_config);
        let (client, conn) = tokio_postgres::connect(config, tls)
            .await
            .map_err(|e| ConnectionError::BadConnection(e.to_string()))?;
        tokio::spawn(async move {
            if let Err(e) = conn.await {
                eprintln!("Database connection: {e}");
            }
        });
        diesel_async::async_connection_wrapper::AsyncConnectionWrapper::establish(
            AsyncPgConnection::try_from(client).await.unwrap(),
        )
    };
    fut.boxed()
}

fn root_certs() -> rustls::RootCertStore {
    let mut roots = rustls::RootCertStore::empty();
    let certs = rustls_native_certs::load_native_certs().expect("Certs not loadable!");
    let certs: Vec<_> = certs.into_iter().map(|cert| cert.0).collect();
    roots.add_parsable_certificates(&certs);
    roots
}
```

**Reply by weiznich** [reply, marked as answer]
Thanks for providing these information. I'm not sure how you've got your `table!` definition, but it looks like you have written that on your own as the table schema is set as part of your `table!` definition. If you use `diesel print-schema` to generate it, it will automatically add the right schema there.

If you manually write your `table!` definition you need to include the table schema like that:

```rust
table! {
    your_schema.activity_history (id) {
        id -> BigInt,
        timestamp -> Timestamp,
        reporter -> Text,
        activity -> Text,
        data -> Json,
    }
}
```

This generates internally the following module structure:

```
pub mod your_schema {
     pub mod activity_history {
         pub struct table;
         // for more content see this guide:
         // https://diesel.rs/guides/schema-in-depth.html
     }
}
```

Now `#[diesel(table_name = …)]` requires access to module named like the table. How you import that module is up to you, so you could either do a `use crate::schema::your_schema::activity_history;` in combination with `#[diesel(table_name = activity_history)]` or you could use the full path in the attribute like #[diesel(table_name =  crate::schema::your_schema::activity_history)]` or you could import parts of the path via use and put other parts in the attribute.

**Reply by Abdillah** [reply]
>     your_schema.activity_history (id) {

Ah, that's it!

I wrote my own `table!` definition, because I don't plan on writing migration (the table is already created in the server).

**Reply by weiznich** [reply]
In that case consider using `diesel print-schema` to get the `table!` definitions form your existing database.

**Reply by Abdillah** [reply]
Okay, thanks for the help
